staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go

```diff
@@ -0,0 +1,575 @@
+		dm.mergedDiscoveryHandler.AddGroupVersion(gv.Group, apidiscoveryv2beta1.APIVersionDiscovery{
+			Version: gv.Version,
+		})
+	}
+
+	return nil
+}
+
+// Spwans a goroutune which waits for added/updated apiservices and updates
+// the discovery document accordingly
+func (dm *discoveryManager) Run(stopCh <-chan struct{}) {
+	klog.Info("Starting ResourceDiscoveryManager")
+
+	// Shutdown the queue since stopCh was signalled
+	defer dm.dirtyAPIServiceQueue.ShutDown()
+
+	// Spawn workers
+	// These workers wait for APIServices to be marked dirty.
+	// Worker ensures the cached discovery document hosted by the ServiceReference of
+	// the APIService is at least as fresh as the APIService, then includes the
+	// APIService's groupversion into the merged document
+	for i := 0; i < 2; i++ {
+		go func() {
+			for {
+				next, shutdown := dm.dirtyAPIServiceQueue.Get()
+				if shutdown {
+					return
+				}
+
+				func() {
+					defer dm.dirtyAPIServiceQueue.Done(next)
+
+					if err := dm.syncAPIService(next.(string)); err != nil {
+						dm.dirtyAPIServiceQueue.AddRateLimited(next)
+					} else {
+						dm.dirtyAPIServiceQueue.Forget(next)
+					}
+				}()
+			}
+		}()
+	}
+
+	// Ensure that apiregistration.k8s.io is the first group in the discovery group.
+	dm.mergedDiscoveryHandler.SetGroupPriority(APIRegistrationGroup, APIRegistrationGroupPriority)
+
+	wait.PollUntil(1*time.Minute, func() (done bool, err error) {
+		dm.servicesLock.Lock()
+		defer dm.servicesLock.Unlock()
+
+		now := time.Now()
+
+		// Mark all non-local APIServices as dirty
+		for key, info := range dm.apiServices {
+			info.lastMarkedDirty = now
+			dm.apiServices[key] = info
+			dm.dirtyAPIServiceQueue.Add(key)
+		}
+		return false, nil
+	}, stopCh)
+}
+
+// Adds an APIService to be tracked by the discovery manager. If the APIService
+// is already known
+func (dm *discoveryManager) AddAPIService(apiService *apiregistrationv1.APIService, handler http.Handler) {
+	dm.servicesLock.Lock()
+	defer dm.servicesLock.Unlock()
+
+	// If service is nil then its information is contained by a local APIService which is has already been added to the manager.
+	if apiService.Spec.Service == nil {
+		// The legacy group is a special case and is not served by the /apis handler
+		if apiService.Spec.Group == "" {
+			return
+		}
+		// Only add the priority defined in the APIService for sorting the discovery document.
+		dm.mergedDiscoveryHandler.SetGroupPriority(apiService.Spec.Group, int(apiService.Spec.GroupPriorityMinimum))
+		return
+	}
+
+	// Add or update APIService record and mark it as dirty
+	dm.setInfoForAPIService(apiService.Name, &groupVersionInfo{
+		groupPriority:   int(apiService.Spec.GroupPriorityMinimum),
+		handler:         handler,
+		lastMarkedDirty: time.Now(),
+		service:         newServiceKey(*apiService.Spec.Service),
+	})
+	dm.dirtyAPIServiceQueue.Add(apiService.Name)
+}
+
+func (dm *discoveryManager) RemoveAPIService(apiServiceName string) {
```

Comment:
Adding code to skip LocalAPIServices will make this code safer for future refactors

Reply:
> Adding code to skip LocalAPIServices will make this code safer for future refactors

re the question defering: this is a small enough change, I suggest going ahead and making it.